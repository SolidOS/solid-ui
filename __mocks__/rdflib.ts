export const Util =  {
  mediaTypeClass: sym
}

export function sym () {
  return {
    dir: sym,
    uri: 'uri',
    value: '',
    elements: [],
    doc: () => {
      return {
        equals: () => {},
        value: ''
      }
    },
    sameTerm: () => false
  }  
}  

export const uri = {
  join: () => {}
}
export function st() {
}

export function literal() {
}
export function graph() {
  return {
    any: () => {
      return sym()
    },
    each: () => {
      return []
    },
    match: () => {
      return []
    },
    query: () => {
      return []
    },
    fetcher: {
      load: () => {},
      nowOrWhenFetched: () => Promise.resolve()
    },
    findTypeURIs: () => {
      return []
    },
    findSuperClassesNT: () => {
      return []
    },
    bottomTypeURIs: () => {
      return []
    },
    statementsMatching: () => {
      return []
    },
    sym
  }
}
export function fetcher() {
  return {}
}
export class Fetcher {
  requested: any
  constructor() {
    this.requested = {}
  }
  load () {
  }
}
export function namedNode() {
  return {}
}
export function NamedNode() {
}
export function Namespace() {
  return () => {}
}
export class UpdateManager {
  editable() {}
  update() {}
}
