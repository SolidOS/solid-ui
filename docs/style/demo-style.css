        input {
            background-color: #eef;
            padding: 0.5em;
            border: .5em solid white;
            font-size: 120%;
        }

        table {

            border-collapse: collapse;
            border: 0.1em solid gray;
            margin: 1em;

        }

        tr.naviMenu {
            background-color: white;
        }

        tr.naviMenu td {
            text-align: middle;
            vertical-align: middle;
            padding-top: 4em;
        }

        table tbody tr td {
            padding: 1em;
            border: 0.1em solid gray;
        }

        td {
            vertical-align: top;
        }

        td.input {
            border-top: 0.1em solid gray;
            border-right: 0.1em dotted gray;
            padding: 1em;
        }

        td.output {
            border-top: 0.1em solid gray;
            padding: 1em;
        }

         td.input pre { padding: 0.5em;}
