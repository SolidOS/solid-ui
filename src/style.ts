// Common readable consistent stylesheet
// to avoid using style sheets which are document-global
// and make programmable style toggling with selection, drag over, etc easier

// These must all end with semicolon so they can be appended to.

const styles = {
  textInputStyle: 'background-color: #eef; padding: 0.5em;  border: .5em solid white; font-size: 100%;',
  buttonStyle: 'background-color: #fff; padding: 0.5em;  border: .01em solid white; font-size: 100%;', // 'background-color: #eef;
// The width of the text field must bot be 100% or it switches to overlapping
  messageBodyStyle: 'white-space: pre-wrap; width: 99%; font-size:100%; border: 0.07em solid #eee; padding: .3em 0.5em; margin: 0.1em;',
  pendingeditModifier: 'color: #bbb;'

}
export default styles
