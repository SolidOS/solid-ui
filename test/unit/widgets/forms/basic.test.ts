import { silenceDebugMessages } from '../../../helpers/setup'
import { lit, namedNode } from 'rdflib'
import {
  fieldLabel,
  fieldStore,
  basicField
} from '../../../../src/widgets/forms/basic'
import { store, ns } from '../../../../src/'
import { textInputStyle } from '../../../../src/style'
import { clearStore } from '../../helpers/clearStore'

silenceDebugMessages()
afterEach(clearStore)

describe('fieldLabel', () => {
  it('exists', () => {
    expect(fieldLabel).toBeInstanceOf(Function)
  })
  const predicates = [
    ns.ui('label'),
    ns.link('message'),
    ns.vcard('fn'),
    ns.foaf('name'),
    ns.dct('title'),
    ns.dc('title'),
    ns.rss('title'),
    ns.contact('fullName'),
    namedNode('http://www.w3.org/2001/04/roadmap/org#name'),
    ns.cal('summary'),
    ns.foaf('nick'),
    ns.rdfs('label')
  ]
  for (let i = 0; i < predicates.length; i++) {
    it(`checks for ${predicates[i]}`, () => {
      for (let j = i; j < predicates.length; j++) {
        store.add(namedNode('http://example.com/#this'), predicates[j], `from ${predicates[j]}`, namedNode('http://example.com/'))
      }
      const anchorElement = fieldLabel(document, namedNode('http://example.com/#this'), namedNode('http://example.com/#this'))
      const firstChar = (i === 0 ? 'f' : 'F')
      const expectedLabel = `${firstChar}rom &lt;${predicates[i].uri}&gt;`
      expect((anchorElement as HTMLElement).innerHTML).toEqual(expectedLabel)
    })
  }
  it('sets the property as the href', () => {
    store.add(namedNode('http://example.com/#this'), predicates[0], `from ${predicates[0]}`, namedNode('http://example.com/'))
    const anchorElement = fieldLabel(document, namedNode('http://example.com/#the-property'), namedNode('http://example.com/#this'))
    const expectedHref = 'http://example.com/#the-property'
    expect((anchorElement as HTMLAnchorElement).href).toEqual(expectedHref)
  })
  it('sets the style', () => {
    store.add(namedNode('http://example.com/#this'), predicates[0], `from ${predicates[0]}`, namedNode('http://example.com/'))
    const anchorElement = fieldLabel(document, namedNode('http://example.com/#the-property'), namedNode('http://example.com/#this'))
    const expectedStyle = 'color: #3B5998; text-decoration: none;'
    expect((anchorElement as HTMLAnchorElement).getAttribute('style')).toEqual(expectedStyle)
  })
  it('shows an error if property is undefined', () => {
    store.add(namedNode('http://example.com/#this'), predicates[0], `from ${predicates[0]}`, namedNode('http://example.com/'))
    const textNode = fieldLabel(document, undefined, namedNode('http://example.com/#this'))
    const expectedText = '@@Internal error: undefined property'
    expect((textNode as Text).wholeText).toEqual(expectedText)
  })
})

describe('fieldStore', () => {
  it('exists', () => {
    expect(fieldStore).toBeInstanceOf(Object)
  })
  it('returns the first matching why, if editable', () => {
    store.add(namedNode('http://example.com/#s'), namedNode('http://example.com/#p'), namedNode('http://example.com/#o1'), namedNode('http://why1.com/'))
    store.add(namedNode('http://example.com/#s'), namedNode('http://example.com/#p'), namedNode('http://example.com/#o2'), namedNode('http://why2.com/'))

    expect(fieldStore(namedNode('http://example.com/#s'), namedNode('http://example.com/#p'), namedNode('http://example.com/#def'))).toEqual(namedNode('http://why1.com/'))
  })
  // TODO: Cannot set literal as why parameter with new types
  it.skip('returns def if the first matching why is not editable', () => {
    store.add(namedNode('http://example.com/#s'), namedNode('http://example.com/#p'), namedNode('http://example.com/#o1'), lit('some literal 1'))
    store.add(namedNode('http://example.com/#s'), namedNode('http://example.com/#p'), namedNode('http://example.com/#o2'), lit('some literal 2'))

    expect(fieldStore(namedNode('http://example.com/#s'), namedNode('http://example.com/#p'), namedNode('http://def.com/'))).toEqual(namedNode('http://def.com/'))
  })
  it('returns def when there is no matching statement', () => {
    expect(fieldStore(namedNode('http://example.com/#s'), namedNode('http://example.com/#p'), namedNode('http://def.com/'))).toEqual(namedNode('http://def.com/'))
  })
})

describe('basicField', () => {
  it('exists', () => {
    expect(fieldStore).toBeInstanceOf(Function)
  })
  it('creates a basic field', () => {
    const container = undefined
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), namedNode('http://example.com/#some-property'), namedNode('http://example.com/'))

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    expect(result).toMatchSnapshot()
  })

  it('adds the basic field to the container, if provided', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), namedNode('http://example.com/#some-property'), namedNode('http://example.com/'))

    basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    expect(container).toMatchSnapshot()
  })

  it('handles keyup events', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '555-1234'
    const event = new Event('keyup')
    inputElement.dispatchEvent(event)
  })

  it('handles change events', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '555-1234'
    const event = new Event('change')
    inputElement.dispatchEvent(event)
  })

  it('goes green if value matches pattern', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = namedNode('http://www.w3.org/ns/ui#PhoneField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '555-1234'
    const event = new Event('keyup')
    inputElement.dispatchEvent(event)
    expect(inputElement.getAttribute('style')).toEqual('background-color: #eef; padding: 0.5em;  border: .05em solid #88c;  border-radius:0.2em; font-size: 100%; margin:0.2em; color: green;')
  })

  it('goes red if value doesnt match pattern', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = namedNode('http://www.w3.org/ns/ui#PhoneField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = 'not a valid phone number'
    const event = new Event('keyup')
    inputElement.dispatchEvent(event)
    expect(inputElement.getAttribute('style')).toEqual('background-color: #eef; padding: 0.5em;  border: .05em solid #88c;  border-radius:0.2em; font-size: 100%; margin:0.2em; color: red;')
  })

  it('handles change if value doesnt match pattern', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = namedNode('http://www.w3.org/ns/ui#PhoneField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = 'not a valid phone number'
    const event = new Event('change')
    inputElement.dispatchEvent(event)
  })

  it('calls updater on change', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('http://example.com/#initial-value'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = 'changed value'
    const event = new Event('change')
    inputElement.dispatchEvent(event)
    expect(store.updater.updated).toEqual(true)
  })

  it('calls updater on change for a NamedNodeUriField', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = ns.ui('NamedNodeURIField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(subject, property, namedNode('http://example.com/#initial-value'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = 'http://example.com/#changed-value'
    const event = new Event('change')
    inputElement.dispatchEvent(event)
    expect(store.updater.updated).toEqual(true)
  })

  it('calls updater on change for a field with uriPrefix', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = namedNode('http://www.w3.org/ns/ui#PhoneField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(subject, property, namedNode('tel:11111'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '999000'
    const event = new Event('change')
    inputElement.dispatchEvent(event)
    expect(store.updater.updated).toEqual(true)
  })

  it('calls updater on change for a field with dt', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = ns.ui('FloatField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(subject, property, 1.1111, doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '9.99000'
    const event = new Event('change')
    inputElement.dispatchEvent(event)
    expect(store.updater.updated).toEqual(true)
  })

  it('creates triple if missing', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '9.99000'
    const event = new Event('change')
    inputElement.dispatchEvent(event)
    expect(store.updater.updated).toEqual(true)
  })

  it('Can update multiple docs', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = ns.ui('FloatField')
    const property = namedNode('http://example.com/#some-property')
    const doc1 = namedNode('http://example.com/1')
    const doc2 = namedNode('http://example.com/2')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc1)
    store.add(form, ns.rdf('type'), formType, doc1)
    store.add(subject, property, 1.1111, doc1)
    store.add(subject, property, 1.5, doc1)
    store.add(subject, property, 1.1111, doc2)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc1,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '1.1111'
    const event = new Event('change')
    inputElement.dispatchEvent(event)
    expect(store.updater.updated).toEqual(true)
  })

  it('Reports update success', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, 1.1111, doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '1.1111'
    const event = new Event('change')
    store.updater.reportSuccess = true
    inputElement.dispatchEvent(event)
    expect(callbackFunction.mock.calls).toEqual([[true, 'body']])
  })

  it('Reports update failure', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, 1.1111, doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '1.1111'
    const event = new Event('change')
    store.updater.reportSuccess = false
    inputElement.dispatchEvent(event)
    expect(callbackFunction.mock.calls).toEqual([[false, 'body']])
  })

  it('Reports update failure to update multiple docs', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const doc2 = namedNode('http://example.com/2')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, 1.1111, doc)
    store.add(subject, property, 1.1111, doc2)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const inputElement = result.childNodes[1].childNodes[0] as HTMLInputElement
    inputElement.value = '1.1111'
    const event = new Event('change')
    store.updater.reportSuccess = false
    inputElement.dispatchEvent(event)
    expect(callbackFunction.mock.calls).toEqual([[false, 'body']])
  })

  it('Can derive the doc if undefined', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const doc = undefined
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), namedNode('http://example.com/#some-property'), namedNode('http://example.com/'))
    // Let it derive the correct doc from this:
    store.add(subject, namedNode('http://example.com/#some-property'), namedNode('http://example.com/#o1'), namedNode('http://why1.com/'))

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    // From the fact that it's picking up the 'http://example.com/#o1' object
    // and that apparently it didn't error on line
    // https://github.com/solid/solid-ui/blob/660ab45/src/widgets/forms/basic.ts#L135
    // we can conclude that it successfully chose namedNode('http://why1.com/')
    // as the value for `doc`.
    expect((result.childNodes[1].childNodes[0] as HTMLInputElement).value).toEqual('http://example.com/#o1')
  })

  it('Sets the current value from rdflib (Literal)', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, 'current value', doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    expect((result.childNodes[1].childNodes[0] as HTMLInputElement).value).toEqual('current value')
  })

  it('Sets the current value from rdflib (NamedNode)', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('http://example.com/#current-value'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    expect((result.childNodes[1].childNodes[0] as HTMLInputElement).value).toEqual('http://example.com/#current-value')
  })

  it('Can read uriPrefix from fieldParams', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = namedNode('http://www.w3.org/ns/ui#PhoneField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    expect((result.childNodes[1].childNodes[0] as HTMLInputElement).value).toEqual('123412341234')
  })

  it('Can read style from fieldParams', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = ns.ui('FloatField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const style = (result.childNodes[1].childNodes[0] as HTMLInputElement).getAttribute('style')
    expect(style).toEqual('text-align: right')
  })

  it('Defaults to textInputStyle', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const style = (result.childNodes[1].childNodes[0] as HTMLInputElement).getAttribute('style')
    expect(style).toEqual(textInputStyle)
  })

  it('Can read type from fieldParams', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = ns.ui('TimeField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const type = (result.childNodes[1].childNodes[0] as HTMLInputElement).getAttribute('type')
    expect(type).toEqual('time')
  })

  it('Defaults to type text', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const type = (result.childNodes[1].childNodes[0] as HTMLInputElement).getAttribute('type')
    expect(type).toEqual('text')
  })

  it('Form size has precedence over fieldParams size', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = ns.ui('TimeField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(form, ns.ui('property'), property, doc)
    store.add(form, ns.ui('size'), 30, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const size = (result.childNodes[1].childNodes[0] as HTMLInputElement).getAttribute('size')
    expect(size).toEqual('30')
  })

  it('Can read size from fieldParams', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const formType = ns.ui('TimeField')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.rdf('type'), formType, doc)
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const size = (result.childNodes[1].childNodes[0] as HTMLInputElement).getAttribute('size')
    expect(size).toEqual('10')
  })

  it('Defaults to size 20', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const size = (result.childNodes[1].childNodes[0] as HTMLInputElement).getAttribute('size')
    expect(size).toEqual('20')
  })

  it('Can read maxLength from form property', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://example.com/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(form, ns.ui('maxLength'), 100, doc)
    store.add(subject, property, namedNode('tel:123412341234'), doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    const maxLength = (result.childNodes[1].childNodes[0] as HTMLInputElement).getAttribute('maxLength')
    expect(maxLength).toEqual('100')
  })

  it('Shows as readOnly if not editable', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const property = namedNode('http://example.com/#some-property')
    const doc = namedNode('http://not.editable/')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    store.add(form, ns.ui('property'), property, doc)
    store.add(subject, property, 'set in stone', doc)

    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      doc,
      callbackFunction
    )
    expect((result.childNodes[1].childNodes[0] as HTMLInputElement).readOnly).toEqual(true)
  })

  it('shows error if no property given', () => {
    const container = document.createElement('div')
    const already = {}
    const subject = namedNode('http://example.com/#this')
    const form = namedNode('http://example.com/#form')
    const store = namedNode('http://example.com/#store')
    const callbackFunction = jest.fn() // TODO: https://github.com/solid/solid-ui/issues/263
    const result = basicField(
      document,
      container,
      already,
      subject,
      form,
      store,
      callbackFunction
    )
    expect(result).toMatchSnapshot()
  })
})
