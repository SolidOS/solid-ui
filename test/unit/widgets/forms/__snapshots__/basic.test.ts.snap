// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basicField adds the basic field to the container, if provided 1`] = `
<div>
  <tr>
    <td
      class="formFieldName"
      style="  vertical-align: middle;"
    >
      <a
        href="http://example.com/#some-property"
        style="color: #3B5998; text-decoration: none;"
      >
        Some property
      </a>
    </td>
    <td
      class="formFieldValue"
    >
      <input
        maxlength="4096"
        size="20"
        style="background-color: #eef; padding: 0.5em;  border: .05em solid #88c;  border-radius:0.2em; font-size: 100%; margin:0.2em; "
        type="text"
      />
    </td>
  </tr>
</div>
`;

exports[`basicField creates a basic field 1`] = `
<tr>
  <td
    class="formFieldName"
    style="  vertical-align: middle;"
  >
    <a
      href="http://example.com/#some-property"
      style="color: #3B5998; text-decoration: none;"
    >
      Some property
    </a>
  </td>
  <td
    class="formFieldValue"
  >
    <input
      maxlength="4096"
      size="20"
      style="background-color: #eef; padding: 0.5em;  border: .05em solid #88c;  border-radius:0.2em; font-size: 100%; margin:0.2em; "
      type="text"
    />
  </td>
</tr>
`;

exports[`basicField shows error if no property given 1`] = `
<tr>
  <td
    class="formFieldName"
    style="  vertical-align: middle;"
  />
  <td
    class="formFieldValue"
  />
  Error: No property given for text field: &lt;http://example.com/#form&gt;
</tr>
`;
